services/horizon: Move metrics endpoint to new actions code

2opremio opened the pull request:
<!-- If you're making a doc PR or something tiny where the below is irrelevant, delete this
template and use a short description, but in your description aim to include both what the
change is, and why it is being made, with enough context for anyone to understand. -->

<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Move metrics endpoint to new actions code as part of #2582 (picked an _easy_ endpoint to start with, so that I can collect early feedback and learn the mechanics quickly)

### Why

The old actions code is awkward and convoluted.

### Known limitations

N/A

2opremio commented on services/horizon/internal/handler.go:
```diff
@@ -591,3 +592,25 @@ func buildPage(r *http.Request, records []hal.Pageable) (hal.Page, error) {
 	return page, nil
 }
+
+type metricsAction interface {
+	PrometheusFormat(w io.Writer) error
+	GetMetricsSnapshot() map[string]interface{}
+}
```
I am not sure this interface actually helps, we could simply use `MetricsAction` directly

bartekn commented:
It's going into a right direction but I should have explained how new actions work in https://github.com/stellar/go/issues/2582 - sorry!

To remove code duplication we decided that actions should implement one of the methods below (depends on an action purpose):
```go
// For simple one-object responses:
GetResource(w actions.HeaderWriter, r *http.Request) (hal.Pageable, error)
// For page of objects:
GetResourcePage(w actions.HeaderWriter, r *http.Request) ([]hal.Pageable, error)
// For streamable endpoints
GetResource(actions.HeaderWriter, r *http.Request) (actions.StreamableObjectResponse, error)
```
So when you create a new action you simply create a handler struct that implements one of the interfaces above. See examples in `actions` package, ex. `GetOrderbookHandler`.

So to move `/metrics` we need to:
1. Update `protocols/horizon/Metrics` struct - not sure why it's not up to date and not used (probably part of the mess in old actions package). It probably doesn't implement `hal.Pageable` so we probably need to implement it (can return an empty string). The reason we decided to use `hal.Pageable` is because it was suitable for majority of endpoints so instead of inventing something tricky we were ok with having to implement `Pageable` for something doesn't really implement it - like here.
2. Next, we need `GetMetricsHandler` - I think you already have it but we should use similar naming as existing handlers.
3. Finally, we need to update router configuration to use the new action (you can use `objectActionHandler` helper struct to create a HTTP handler responsible for rendering).

With above in place you don't need to care about rendering, a shared code responsible for rendering `hal.Pageable` will do it for you.

Now, `/metrics` is a bit tricky because it can also return prometheus format response. We could create a new interface for prometheus responses and update entire stack with it but I think it does not make much sense because `/metrics` is and will be the only place where we need this. What we can do is:
* embed `io.Writer` in `actions.HeaderWriter` interface (we can think about renaming it too because it's not only about headers now...),
* update `objectActionHandler.ServeHTTP` to not render JSON response if returned `hal.Pageable` is `nil`.

This way we can make metrics action write directly to `http.ResponseWriter` in case prometheus format is requested.

2opremio commented:
Thanks a lot for the detailed explanation! I will make the changes and ping you when it's ready to review again.

2opremio commented:
> To remove code duplication we decided that actions should implement one of the methods below (depends on an action purpose)

I am not sure the existing machinery is well suited for the metrics endpoint. For one, we need to bend it in order to support the `metrics` endpoint:

* We need to pass on the http response writer in order to write the prometheus-conformant response.
* We also need to check the `prometheus_format` query parameter, to distinguish

In this particular case, conforming with the abstraction used by the rest of the endpoints seems to hurt legibility more than aid it.

Also, I think it fails the purpose of _remov[ing] code duplication_ , since we need to add code to cover the cornercases.

I am generally in favor of consistency overall, but not in this particular case.

> 1. Update `protocols/horizon.Metrics` struct - not sure why it's not up to date and not used (probably part of the mess in old actions package).

I can do this but using hardcoded code for every metric is going to be a pain without using reflection. Plus, it may be hard to maintain. Wouldn't it be better to simply use maps instead?

Also, who is using the non-prometheus endpoint? (i.e. why is it there? is it worth maintaining?)

bartekn commented:
I think I agree with removing JSON response in this endpoint. Prometheus format is simple enough to parse in case somebody needs it outside Prometheus.

However, I think it's worth embedding `io.Writer` in `HeaderWriter` (explained in my previous comment) because it's probably the easiest we can do to support another format that's used in a single endpoint. Do you have an alternative proposal?

2opremio commented:
I think I have addressed everything:

1. I have removed the JSON variant of `GET /metrics`
2. I have replaced `actions.HeaderWriter` by `http.ReponseWriter` (adding an io.Writer to it would effectively had implied re-inventing the wheel).
3. I have made the metrics code adhere to the `objectAction` interface, modifying the handler to accept a nil page response.

I still don't agree with (2) and (3). I think its leaky (it lets actions bypass the rendering code) and bends the abstraction for a single usecase. Since we need to add an exception for the metrics code, we might as well skip the the abstraction altogether and make an explicit corner case (just like it happened before my last change). Its more explicit (i.e. more readable) and it doesn't polute other handlers.

I would like to know @tamirms 's take on this one.

tamirms commented:
> I still don't agree with (2) and (3). I think its leaky (it lets actions bypass the rendering code) and bends the abstraction for a single usecase. Since we need to add an exception for the metrics code, we might as well skip the the abstraction altogether and make an explicit corner case (just like it happened before my last change). Its more explicit (i.e. more readable) and it doesn't polute other handlers.

@2opremio yeah, I completely agree. I never really liked the abstraction to begin with

bartekn commented:
> I still don't agree with (2) and (3). I think its leaky (it lets actions bypass the rendering code) and bends the abstraction for a single usecase. Since we need to add an exception for the metrics code, we might as well skip the the abstraction altogether and make an explicit corner case (just like it happened before my last change). Its more explicit (i.e. more readable) and it doesn't polute other handlers.

OK, I agree that this is more readable.

> @2opremio yeah, I completely agree. I never really liked the abstraction to begin with

@tamirms do you mean (2) and (3) or the new actions architecture as a whole?

tamirms commented:
@bartekn I would prefer to use the standard http handler interface wherever possible. If an endpoint doesn't fit well within the `objectAction` or `pageableAction` interface, rather than expand the scope of the interface I would prefer to implement the endpoint as a standard http handler

bartekn commented:
OK, this makes a lot of sense when we no longer return JSON response for `/metrics`.

tamirms commented on services/horizon/internal/actions/metrics.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+// MetricsAction is the action handler for the /metrics endpoint
+type MetricsAction struct {
```
In the other files we use a suffix of "Handler" instead of "Action"

2opremio replied:
I will change it, thanks.

tamirms commented on services/horizon/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ file. This project adheres to [Semantic Versioning](http://semver.org/).
 ## Unreleased
 
 * Replace `SequenceProvider` implementation with one which queries the Horizon DB for sequence numbers instead of the Stellar Core DB.
+* Remove JSON variant of `GET /metrics`, both in the server and client code.
```
can you also update the changelog for horizonclient?

bartekn replied:
```suggestion
* Remove JSON variant of `GET /metrics`, both in the server and client code. It's using Prometheus format by default now.
```

bartekn commented on protocols/horizon/main.go:
```diff
@@ -646,33 +646,6 @@ type LogTotalMetric struct {
-	LoggingPanic                   LogMetric      `json:"logging.panic"`
-	LoggingWarning                 LogMetric      `json:"logging.warning"`
-	RequestsFailed                 LogMetric      `json:"requests.failed"`
-	RequestsSucceeded              LogMetric      `json:"requests.succeeded"`
-	RequestsTotal                  LogTotalMetric `json:"requests.total"`
-	CoreLatestLedger               SingleMetric   `json:"stellar_core.latest_ledger"`
```
These structs can be removed too.

bartekn commented on services/horizon/internal/handler.go:
```diff
@@ -605,3 +606,19 @@ func buildPage(r *http.Request, records []hal.Pageable) (hal.Page, error) {
+
+func HandleMetrics(action metricsAction) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("prometheus_format") == "true" {
+			if err := action.PrometheusFormat(w); err != nil {
+				problem.Render(r.Context(), w, err)
+			}
+			return
+		}
```
Is this code needed? I thought that prometheus format is returned as default right now.

2opremio replied:
I wanted to keep that part backwards compatible. But it probably makes sense. Will fix.
